File: src/main/java/org/ant/plugin/Execute.java
package org.ant.plugin;

import com.mojang.brigadier.Command;
import com.mojang.brigadier.context.CommandContext;
import io.papermc.paper.command.brigadier.CommandSourceStack;
import org.bukkit.Location;
import org.bukkit.entity.Entity;

import java.util.Optional;

public class Execute {
    public static int set_board(CommandContext<CommandSourceStack> ctx, String mode) {
        Entity executor = ctx.getSource().getExecutor();
        if(executor != null){
            Location location = executor.getLocation().getBlock().getLocation();
            String name = ctx.getArgument("name", String.class);
            switch (mode) {
                case "chess":
                    Game.getInstance().chess_games.put(name, new Chess(location));
                    break;
                case "gomoku":
                    Game.getInstance().gomoku_games.put(name, new Gomoku(location, Optional.empty(), Optional.empty()));
                    break;
                case "reversi":
                    Game.getInstance().reversi_games.put(name, new Reversi(location, Optional.empty(), Optional.empty()));
                    break;
                case "lights_out":
                    int size = ctx.getArgument("size", int.class);
                    Game.getInstance().lightsOut_games.put(name, new LightsOut(location, size, Optional.empty(), Optional.empty()));
                    break;
            }
        }
        return Command.SINGLE_SUCCESS;
    }

    public  static int set_display(CommandContext<CommandSourceStack> ctx, String mode) {
        String name = ctx.getArgument("name", String.class);
        String align = ctx.getArgument("align", String.class);
        Entity executor = ctx.getSource().getExecutor();
        if(executor != null){
            Location location = executor.getLocation().getBlock().getLocation();
            switch (mode) {
                case "gomoku":
                    if(Game.getInstance().gomoku_games.containsKey(name)) {
                        Gomoku gomoku = Game.getInstance().gomoku_games.get(name);
                        gomoku.remove_display();
                        gomoku.set_display(location, align);
                    }
                    break;
                case "reversi":
                    if(Game.getInstance().reversi_games.containsKey(name)) {
                        Reversi reversi = Game.getInstance().reversi_games.get(name);
                        reversi.remove_display();
                        reversi.set_display(location, align);
                    }
                    break;
                case "lights_out":
                    if(Game.getInstance().lightsOut_games.containsKey(name)) {
                        LightsOut lightsOut = Game.getInstance().lightsOut_games.get(name);
                        lightsOut.remove_display();
                        lightsOut.set_display(location, align);
                    }
                    break;
            }
        }
        return Command.SINGLE_SUCCESS;
    }

    public static int reset_board(CommandContext<CommandSourceStack> ctx, String mode) {
        String name = ctx.getArgument("name", String.class);
        switch (mode) {
            case "chess":
                if(Game.getInstance().chess_games.containsKey(name))Game.getInstance().chess_games.get(name).reset();
                break;
            case "gomoku":
                if(Game.getInstance().gomoku_games.containsKey(name))Game.getInstance().gomoku_games.get(name).reset();
                break;
            case "reversi":
                if(Game.getInstance().reversi_games.containsKey(name))Game.getInstance().reversi_games.get(name).reset();
                break;
            case "lights_out":
                if(Game.getInstance().lightsOut_games.containsKey(name))Game.getInstance().lightsOut_games.get(name).reset();
                break;
        }
        return Command.SINGLE_SUCCESS;
    }

    public static int remove_board(CommandContext<CommandSourceStack> ctx, String mode) {
        String name = ctx.getArgument("name", String.class);
        switch (mode) {
            case "chess":
                if(Game.getInstance().chess_games.containsKey(name)) {
                    Game.getInstance().chess_games.get(name).remove();
                    Game.getInstance().chess_games.remove(name);
                }
                break;
            case "gomoku":
                if(Game.getInstance().gomoku_games.containsKey(name)) {
                    Game.getInstance().gomoku_games.get(name).remove();
                    Game.getInstance().gomoku_games.get(name).remove_display();
                    Game.getInstance().gomoku_games.remove(name);
                }
                break;
            case "reversi":
                if(Game.getInstance().reversi_games.containsKey(name)) {
                    Game.getInstance().reversi_games.get(name).remove();
                    Game.getInstance().reversi_games.get(name).remove_display();
                    Game.getInstance().reversi_games.remove(name);
                }
                break;
            case "lights_out":
                if(Game.getInstance().lightsOut_games.containsKey(name)) {
                    Game.getInstance().lightsOut_games.get(name).remove();
                    Game.getInstance().lightsOut_games.get(name).remove_display();
                    Game.getInstance().lightsOut_games.remove(name);
                }
                break;
        }
        return Command.SINGLE_SUCCESS;
    }

    public static int remove_display(CommandContext<CommandSourceStack> ctx, String mode) {
        String name = ctx.getArgument("name", String.class);
        switch (mode) {
            case "gomoku":
                if(Game.getInstance().gomoku_games.containsKey(name))Game.getInstance().gomoku_games.get(name).remove_display();
                break;
            case "reversi":
                if(Game.getInstance().reversi_games.containsKey(name))Game.getInstance().reversi_games.get(name).remove_display();
                break;
            case "lights_out":
                if(Game.getInstance().lightsOut_games.containsKey(name))Game.getInstance().lightsOut_games.get(name).remove_display();
                break;
        }
        return Command.SINGLE_SUCCESS;
    }

    public static int get_player(CommandContext<CommandSourceStack> ctx, String mode) {
        String name = ctx.getArgument("name", String.class);
        switch (mode) {
            case "chess":
                if(Game.getInstance().chess_games.containsKey(name))return Game.getInstance().chess_games.get(name).player + 1;
            case "gomoku":
                if(Game.getInstance().gomoku_games.containsKey(name))return Game.getInstance().gomoku_games.get(name).player;
            case "reversi":
                if(Game.getInstance().reversi_games.containsKey(name))return Game.getInstance().reversi_games.get(name).player;
        }
        return -1;
    }
}

File: src/main/kotlin/org/ant/game/Execute.kt
package org.ant.game

import com.mojang.brigadier.Command
import com.mojang.brigadier.context.CommandContext
import net.minecraft.commands.CommandSourceStack
import org.ant.game.gameimpl.Chess
import org.ant.game.gameimpl.ConnectFour
import org.ant.game.gameimpl.Gomoku
import org.ant.game.gameimpl.LightsOut
import org.ant.game.gameimpl.Reversi
import org.ant.game.gameimpl.ScoreFour

class Execute(private val instance: Game) {
    fun setBoard(ctx: CommandContext<CommandSourceStack>, mode: GameDeSerializable): Int {
        val executor = ctx.source.executor
        if (executor != null) {
            val location = executor.location.block.location
            val name = ctx.getArgument("name", String::class.java)
            when (mode) {
                Chess -> instance.chessGames[name] = Chess(location)
                Gomoku -> instance.gomokuGames[name] = Gomoku(instance, location, null, null)
                Reversi -> instance.reversiGames[name] = Reversi(instance, location, null, null)
                LightsOut -> {
                    val size = ctx.getArgument("size", Int::class.java)
                    instance.lightsOutGames[name] = LightsOut(location, size, null, null)
                }
                ConnectFour -> {
                    val align = ctx.getArgument("align", String::class.java)
                    instance.connectFourGames[name] = ConnectFour(instance, location, align)
                }
                ScoreFour -> instance.scoreFourGames[name] = ScoreFour(instance, location)
            }
        }
        return Command.SINGLE_SUCCESS
    }

    fun setDisplay(ctx: CommandContext<CommandSourceStack>, mode: GameDeSerializable): Int {
        val name = ctx.getArgument("name", String::class.java)
        val align = ctx.getArgument("align", String::class.java)
        val executor = ctx.source.executor
        if (executor != null) {
            val location = executor.location.block.location
            when (mode) {
                Gomoku -> if (instance.gomokuGames.containsKey(name)) {
                    val gomoku = instance.gomokuGames[name]!!
                    gomoku.removeDisplay()
                    gomoku.setDisplay(location, align)
                }
                Reversi -> if (instance.reversiGames.containsKey(name)) {
                    val reversi = instance.reversiGames[name]!!
                    reversi.removeDisplay()
                    reversi.setDisplay(location, align)
                }
                LightsOut -> if (instance.lightsOutGames.containsKey(name)) {
                    val lightsOut = instance.lightsOutGames[name]!!
                    lightsOut.removeDisplay()
                    lightsOut.setDisplay(location, align)
                }
            }
        }
        return Command.SINGLE_SUCCESS
    }

    fun resetBoard(ctx: CommandContext<CommandSourceStack>, mode: GameDeSerializable): Int {
        val name = ctx.getArgument("name", String::class.java)
        when (mode) {
            Chess -> instance.chessGames[name]?.reset()
            Gomoku -> instance.gomokuGames[name]?.reset(null)
            Reversi -> instance.reversiGames[name]?.reset(null)
            LightsOut -> instance.lightsOutGames[name]?.reset(null)
            ConnectFour -> instance.connectFourGames[name]?.reset(null, null)
            ScoreFour -> instance.scoreFourGames[name]?.reset(null, null)
        }
        return Command.SINGLE_SUCCESS
    }

    fun removeBoard(ctx: CommandContext<CommandSourceStack>, mode: GameDeSerializable): Int {
        val name = ctx.getArgument("name", String::class.java)
        when (mode) {
            Chess -> if (instance.chessGames.containsKey(name)) {
                instance.chessGames[name]!!.remove()
                instance.chessGames.remove(name)
            }
            Gomoku -> if (instance.gomokuGames.containsKey(name)) {
                instance.gomokuGames[name]!!.remove()
                instance.gomokuGames[name]!!.removeDisplay()
                instance.gomokuGames.remove(name)
            }
            Reversi -> if (instance.reversiGames.containsKey(name)) {
                instance.reversiGames[name]!!.remove()
                instance.reversiGames[name]!!.removeDisplay()
                instance.reversiGames.remove(name)
            }
            LightsOut -> if (instance.lightsOutGames.containsKey(name)) {
                instance.lightsOutGames[name]!!.remove()
                instance.lightsOutGames[name]!!.removeDisplay()
                instance.lightsOutGames.remove(name)
            }
            ConnectFour -> if (instance.connectFourGames.containsKey(name)) {
                instance.connectFourGames[name]!!.remove()
                instance.connectFourGames.remove(name)
            }
            ScoreFour -> if (instance.scoreFourGames.containsKey(name)) {
                instance.scoreFourGames[name]!!.remove()
                instance.scoreFourGames.remove(name)
            }
        }
        return Command.SINGLE_SUCCESS
    }

    fun removeDisplay(ctx: CommandContext<CommandSourceStack>, mode: GameDeSerializable): Int {
        val name = ctx.getArgument("name", String::class.java)
        when (mode) {
            Gomoku -> if (instance.gomokuGames.containsKey(name)) instance.gomokuGames[name]!!.removeDisplay()
            Reversi -> if (instance.reversiGames.containsKey(name)) instance.reversiGames[name]!!.removeDisplay()
            LightsOut -> if (instance.lightsOutGames.containsKey(name)) instance.lightsOutGames[name]!!.removeDisplay()
        }
        return Command.SINGLE_SUCCESS
    }
}
